modules/auxiliary/gather/enum_dns.rb

```diff
@@ -1,483 +1,420 @@
-        OptAddress.new('NS', [ false, "Specify the nameserver to use for queries (default is system DNS)" ]),
-        OptAddressRange.new('IPRANGE', [false, "The target address range or CIDR identifier"]),
-        OptBool.new('STOP_WLDCRD', [ true, 'Stops bruteforce enumeration if wildcard resolution is detected', false])
+        OptString.new('DOMAIN', [true, 'The target domain']),
+        OptBool.new('ENUM_AXFR', [true, 'Initiate a zone transfer against each NS record', true]),
+        OptBool.new('ENUM_STD', [true, 'Enumerate standard record types (A,MX,NS,TXT and SOA)', true]),
+        OptBool.new('ENUM_BRT', [true, 'Brute force subdomains and hostnames via the supplied wordlist', false]),
+        OptBool.new('ENUM_TLD', [true, 'Perform a TLD expansion by replacing the TLD with the IANA TLD list', false]),
+        OptBool.new('ENUM_SRV', [true, 'Enumerate the most common SRV records', true]),
+        OptBool.new('STOP_WLDCRD', [true, 'Stops bruteforce enumeration if wildcard resolution is detected', false]),
+        OptAddress.new('NS', [false, 'Nameserver for query']),
+        OptInt.new('THREADS', [false, 'Threads for ENUM_BRT', 1]),
+        OptPath.new('WORDLIST', [false, 'Wordlist of subdomains', ::File.join(Msf::Config.data_directory, 'wordlists', 'namelist.txt')])
       ], self.class)
 
     register_advanced_options(
       [
-        OptInt.new('RETRY', [ false, "Number of times to try to resolve a record if no response is received", 2]),
-        OptInt.new('RETRY_INTERVAL', [ false, "Number of seconds to wait before doing a retry", 2]),
-        OptBool.new('TCP_DNS', [false, "Run queries over TCP", false]),
+        OptInt.new('TIMEOUT', [false, 'DNS TIMEOUT', 8]),
+        OptInt.new('RETRY', [false, 'Number of times to try to resolve a record if no response is received', 2]),
+        OptInt.new('RETRY_INTERVAL', [false, 'Number of seconds to wait before doing a retry', 2]),
+        OptBool.new('TCP_DNS', [false, 'Run queries over TCP', false])
       ], self.class)
   end
 
-  def switchdns(target)
-    if not datastore['NS'].nil?
-      print_status("Using DNS Server: #{datastore['NS']}")
-      @res.nameserver=(datastore['NS'])
-      @nsinuse = datastore['NS']
-    else
-      querysoa = @res.query(target, "SOA")
-      if (querysoa)
-        (querysoa.answer.select { |i| i.class == Net::DNS::RR::SOA}).each do |rr|
-          query1soa = @res.search(rr.mname)
-          if (query1soa and query1soa.answer[0])
-            print_status("Setting DNS Server to #{target} NS: #{query1soa.answer[0].address}")
-            @res.nameserver=(query1soa.answer[0].address)
-            @nsinuse = query1soa.answer[0].address
-          end
-        end
-      end
+  def run
+    domain = datastore['DOMAIN']
+    return if domain.blank?
+
+    dns_wildcard(domain) unless datastore['STOP_WLDCRD']
```

Comment:
This does not align with what `STOP_WLDCRD` is documented to do.  The option indicates that it will stop bruteforce enumeration when `true` if a wildcard DNS setup is found, but what it actually does is check for it only if the option is set.  It does not stop enumeration as documented.